### PR TITLE
2165 plugin validation

### LINKED ISSUE #2164: Plugins: validation
## Is your feature request related to a problem? Please describe 👀

Part of epic: #2084 
Two aspects for the plugin validation
- [ ] Sign plugins: provide a way to sign and create a manifest of the plugin files
- [ ] Enforce a minimum version of omSupply 

## Describe the solution you'd like 🎁

for the signing - create a cli action to sign a plugin which creates a manifest and signs it (using a private key?). The server can then check validity of the signature and verify that only the signed files, with correct size and hash are in the plugin directory.

version check: a simple minimum version number that the plugin will require for omSupply

### LINKED ISSUE #2165: Nicely format text for generic patient creation
## Is your feature request related to a problem? Please describe 👀
Gender selection looks very angry...
Screenshot 2023-09-04 at 11 29 07 AM

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #2164

- Add a CLI interface to validate a plugin
- Validates a plugin on the server site
- In dev mode the server ignores validation errors but only logs them

Please review the `service/src/plugin/plugin_validation.md` first for more details.

## Testing Setup
Copy some plugins into the server plugin directory `/app_data/plugins/`. For example, https://drive.google.com/file/d/1kiEhLZFPLr-GSRro-KWxexV6lc2oWy6B/view?usp=sharing

On the frontend you might need to force loading these in dev mode:

```diff
@@ -168,7 +168,7 @@ export const useInitPlugins = () => {
   const { data } = useHost.plugins.list();
 
   useEffect(() => {
-    if (process.env['NODE_ENV'] === 'production')
+    if (process.env['NODE_ENV'] === 'production' || true)
       initRemotePlugins({ addColumnPlugin, addComponentPlugin, data });
     else initLocalPlugins({ addColumnPlugin, addComponentPlugin });
   }, [data, addColumnPlugin, addComponentPlugin]);
```

## Testing
- [ ] Generate self-signing keys as described in the docs
    - [ ] Copy cert over to the `/app_data/plugin_certs` directory
- [ ] Sign a plugin
- [ ] Test that plugin loads without server log warnings (disable cached request in the browser)
- [ ] Remove cert from server, no there should be a warning in the logs
- [ ] Add back the cert but remove the manifest file etc...

Do a similar test for plugins signed by a trusted party (see docs): 
- [ ] Generate mSupply key pair and copy "master" cert into the plugin_certs dir
- [ ] Sign the 3rd party key pair using the "master" keys
- [ ] Test as above

### CODE REVIEW

File: server/cli/Cargo.toml [hunk 1]
```diff
@@ -19,17 +19,17 @@ service = { path = "../service" }
...
+log = "0.4.14"
 reqwest = { version = "0.11.10" }
-tokio = { version = "1.17.0", features = ["macros"] }
+serde = "1.0.126"
+serde_json = "1.0.66"
 simple-log = { version = "1.6" }
-log = "0.4.14"
-async-graphql = "3.0.35"
-anyhow = "1.0.56"
+tokio = { version = "1.17.0", features = ["macros"] }
```
Comment (clemens-msupply) [inline server/cli/Cargo.toml:32]: Just ordered them alphabetically but in the end didn't added something...

File: server/service/src/plugin/validation.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+    fn reload(&mut self) -> anyhow::Result<()> {
+        let certs = load_trusted_certs_from_dir(&self.trusted_cert_path)?;
+
+        self.manifests.clear();
+        let walker = WalkDir::new(&self.plugin_dir).max_depth(1);
+        for entry in walker {
+            let entry = entry?;
+
+            // be conservative and take manifest the timestamp before validating the plugin
```
Comment (mark-prins) [inline server/service/src/plugin/validation.rs:88]: 🤔  Unsure what the intention was - something like this?

```suggestion
            // be conservative and check the manifest before validating the plugin
```

File: server/service/src/plugin/plugin_validation.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+```bash
+openssl req -x509 -newkey rsa:2048 -keyout private.pem -out public.pem -nodes
+```
+
+In this example, the `public.pem` file needs to be copied to the `app_data/plugin_certs` server directory.
+
+## Certificates that are signed by a trusted party
+
+A 3rd party plugin developer can generate their own keypair and then let sign their certificate by mSupply:
```
Comment (mark-prins) [inline server/service/src/plugin/plugin_validation.md:39]:
```suggestion
A 3rd party plugin developer can generate their own private key & signing request and then let mSupply generate a public key. The public key is then added to the list of valid decryption keys:
```

File: server/service/src/plugin/plugin_validation.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+2. mSupply signs the request using their own key-pair like:
+
+```bash
+openssl x509 -req -in public_3rd.csr -CA public.pem -CAkey private.pem -out public_3rd.pem -sha256
+```
+
+The `public_3rd.pem` file is sent back to the 3rd plugin developer.
+
+3. The 3rd developer can now sign their plugin using `private_3rd.pem`and`public_3rd.pem`.
```
Comment (mark-prins) [inline server/service/src/plugin/plugin_validation.md:57]:
```suggestion
3. The 3rd developer can now sign their plugin using `private_3rd.pem` and `public_3rd.pem`.
```

**Review by mark-prins — APPROVED** [review]
🎉  That is fantastic!

A great solution, well thought through and nicely documented. Have tested various scenarios and that's working well

**Comment by clemens-msupply** [comment]
One thought about self-signed certificates. If you can trick somebody to install a doggy plugin you can also trick somebody to install their trusted certificate. So not sure if this is actually a good feature...

File: server/service/src/plugin/validation.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/service/src/plugin/validation.rs:88]: Thanks I clearified it and rearranged the code a bit to make it more clear:

```rust
            // Be conservative and record the manifest timestamp before validating the plugin.
            // For example, when the plugin changes while validating it, the older timestamp would
            // trigger a reload when fetching a plugin (in validate_plugin()).
            let metadata = File::open(&manifest_path)?.metadata()?;
            let manifest_datetime = metadata.modified()?;
```

File: server/service/src/plugin/plugin_validation.md (hunk 3 above)
Comment (clemens-msupply) [reply server/service/src/plugin/plugin_validation.md:39]: Thanks, giving some more details now and changed it to:
```
A 3rd party plugin developer can generate their own private key & signing request and then let mSupply generate a public certificate.
This 3rd party certificate can be used to sign the 3rd party plugin but doesn't need to be added to the remote server.
However, the remote server needs to have the mSupply certificate installed (the one used to sign the 3rd party certificate).
I.e., the mSupply "root" certificate is used to validates the 3rd party certificate shipped with the plugin which in turn is used to validate that the plugin hasn't been modified.

```

File: server/service/src/plugin/plugin_validation.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+# Plugin Signing
+
+The CLI interface can then be used sign a plugin.
+This will place the `manifest.json` and the `manifest.signature` file into the plugin directory.
+For example:
+
+```bash
+cargo run --profile fast --bin remote_server_cli -- sign-plugin -p ./app_data/plugins/StockDonor/ -k pathtokeys/private.pem -c pathtokeyskey/public.pem
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/plugin/plugin_validation.md:66]: Sorry this is in my local config, I removed it...

**Comment by clemens-msupply** [comment]
thanks for the feedback!

**Pull request merged**